### PR TITLE
HDDS-10373. Implement framework for capturing Merkle Tree Metrics.

### PR DESCRIPTION
**Pull request by aswinshakil**
## What changes were proposed in this pull request?
Implement metrics for container reconciliation. Metrics added as of this PR

- Read Latency and Failure
- Write Latency and Failure
- Update Latency and Failure

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10373

## How was this patch tested?

Added metrics to existing tests.

### CODE REVIEW

**Comment by errose28** [comment]
Just moving to draft until #6778 is merged since this builds on top of that.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java [hunk 1]
```diff
@@ -65,8 +70,13 @@ public void writeContainerDataTree(KeyValueContainerData data, ContainerMerkleTr
       ContainerProtos.ContainerChecksumInfo newChecksumInfo = read(data).toBuilder()
           .setDataMerkleTree(tree.toProto())
           .build();
-      write(data, newChecksumInfo);
+      captureLatencyNs(metrics.getWriteContainerMerkleTreeLatencyNS(),
+          () -> write(data, newChecksumInfo));
```
Comment (errose28) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java:74]: Should we also include the time to compute the checksums in this metric? That happens in `tree.toProto()` on line 71.

Comment (aswinshakil) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java:74]: Yes we should. I'll make sure to add them shortly.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java [hunk 2]
```diff
@@ -44,12 +46,15 @@ public class ContainerChecksumTreeManager {
...
   // Each container ID is mapped to a stripe.
   private final Striped<ReadWriteLock> fileLock;
+  private final ContainerMerkleTreeMetrics metrics;
 
   /**
    * Creates one instance that should be used to coordinate all container checksum info within a datanode.
    */
   public ContainerChecksumTreeManager(DatanodeConfiguration dnConf) {
     fileLock = SimpleStriped.readWriteLock(dnConf.getContainerChecksumLockStripes(), true);
+    // TODO: TO unregister metrics on stop.
```
Comment (errose28) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java:56]: I made a note of this [here](https://github.com/apache/ozone/pull/6875/files#r1688351623). This PR connects the manager to `OzoneContainer` where we can unregister the metrics.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java [hunk 3]
```diff
@@ -157,6 +173,14 @@ private void write(KeyValueContainerData data, ContainerProtos.ContainerChecksum
     }
   }
 
+  public File getContainerChecksumFile(KeyValueContainerData data) {
+    return new File(data.getMetadataPath(), data.getContainerID() + ".tree");
+  }
+
+  public ContainerMerkleTreeMetrics getMetrics() {
```
Comment (errose28) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java:180]:
```suggestion
  @VisibleForTesting
  public ContainerMerkleTreeMetrics getMetrics() {
```
This is only needed for test code right?

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java [hunk 4]
```diff
@@ -146,6 +156,12 @@ private ContainerProtos.ContainerChecksumInfo read(KeyValueContainerData data) t
     }
   }
 
+  private ContainerProtos.ContainerChecksumInfo readFile(File checksumFile) throws IOException {
```
Comment (errose28) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java:159]: This method is unused, we can remove it.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java [hunk 5]
```diff
@@ -92,28 +104,26 @@ public void markBlocksAsDeleted(KeyValueContainerData data, SortedSet<Long> dele
           .clearDeletedBlocks()
           .addAllDeletedBlocks(sortedDeletedBlockIDs)
           .build();
-      write(data, checksumInfoBuilder.build());
+      captureLatencyNs(metrics.getUpdateContainerMerkleTreeLatencyNS(),
+          () -> write(data, checksumInfoBuilder.build()));
```
Comment (errose28) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java:108]: Can we move the read and write metrics inside the read and write methods so we don't need to duplicate this code every time they are called?

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java [hunk 6]
```diff
@@ -92,28 +104,26 @@ public void markBlocksAsDeleted(KeyValueContainerData data, SortedSet<Long> dele
...
+      metrics.incrementMerkleTreeUpdateFailures();
+      throw new IOException("Error occurred when updating container merkle tree for containerID "
+          + data.getContainerID(), ex);
     } finally {
       writeLock.unlock();
     }
   }
 
-  public ContainerDiff diff(KeyValueContainerData thisContainer, File otherContainerTree)
+  public ContainerDiff diff(KeyValueContainerData thisContainer, ContainerProtos.ContainerChecksumInfo otherInfo)
```
Comment (errose28) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java:119]: Let's leave this interface unchanged for now. The current plan is that diff will be passed whatever file or bytestream is obtained from the wire and deserialization into a proto will happen within this class.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerMerkleTreeMetrics.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  private MutableRate merkleTreeUpdateLatencyNS;
+
+  @Metric(about = "Merkle tree creation latency")
+  private MutableRate merkleTreeCreateLatencyNS;
+
+  public void incrementMerkleTreeWriteFailures() {
+    this.numMerkleTreeWriteFailure.incr();
+  }
+
+  public void incrementMerkleTreeReadFailures() {
```
Comment (errose28) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerMerkleTreeMetrics.java:68]: This currently unused. Looks like it still needs to be connected.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerMerkleTreeMetrics.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    this.numMerkleTreeReadFailure.incr();
+  }
+
+  public void incrementMerkleTreeUpdateFailures() {
```
Comment (errose28) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerMerkleTreeMetrics.java:72]: Can we get rid of this and just use write failures? Right now "write failures" is incremented for tree updates, but "update failures" is incremented for block delete addition failures, which is not intuitive with their names.

Comment (errose28) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerMerkleTreeMetrics.java:72]: If we want to track block delete updates separately, we can also just use a more intuitive name.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java (hunk 2 above)
Comment (aswinshakil) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java:56]: Yes we need to add it there

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java (hunk 3 above)
Comment (aswinshakil) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java:180]: Right now we are using it for just test. In future we may use it so kept it as it is, but I'll add `@VisibleForTesting`. We can remove it when needed.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerMerkleTreeMetrics.java (hunk 8 above)
Comment (aswinshakil) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerMerkleTreeMetrics.java:72]: We can just have write failure, as both are just writing the file.

**Review by errose28 — APPROVED** [review]
Thanks for the updates @aswinshakil

**Pull request merged**